mava/wrappers/jumanji.py

```diff
@@ -372,3 +379,117 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
+                shaped (num_agents, global_info) if has_global_state = True, None otherwise.
+        """
+        global_observation = None
+        # N: number of nodes, same as _num_customers + 1
+        # V: number of vehicles, same as num_agents
+        # Nodes are composed of (x, y, demands)
+        # Windows are composed of (start_time, end_time)
+        # Coeffs are composed of (early, late)
+        # Vehicles have ((x, y), local_time, capacity)
+
+        # Tuple[(N, 3) : Nodes, (N, 2) : Windows, (N, 2) : Coeffs]
+        customers_info, _ = tree_util.tree_flatten(
+            (observation.nodes, observation.windows, observation.coeffs)
+        )
+        # Tuple[(V, 2) : Coordinates, (V, 1) : Local time, (V, 1) : Capacity]
+        vehicles_info, _ = tree_util.tree_flatten(observation.vehicles)
+
+        # (N * 7, ) : N * (7 : Nodes (3) + Windows (2) + Coeffs (2))
+        customers_info = jnp.column_stack(customers_info).ravel()
+        # (V, 4) : V, (4 : Coordinates (2), Local Time (1), Coeffs (1))
+        vehicles_info = jnp.column_stack(vehicles_info)
+
+        if self.has_global_state:
+            # (V * 4 + N * 7, )
+            global_observation = jnp.concatenate((vehicles_info.ravel(), customers_info))
+            # (V, N * 7 + V * 4)
+            global_observation = jnp.tile(global_observation, (self.num_agents, 1))
+
+        # (V, N * 7)
+        customers_info = jnp.tile(customers_info, (self.num_agents, 1))
+        # (V, 4 + N * 7)
+        observations = jnp.column_stack((vehicles_info, customers_info))
+        return observations, global_observation
+
+    def observation_spec(self) -> specs.Spec[Observation]:
+        step_count = specs.BoundedArray(
+            (self.num_agents,), jnp.int32, 0, self.num_customers + 1, "step_count"
+        )
+        action_mask = specs.BoundedArray(
+            (self.num_agents, self.num_customers + 1), bool, False, True, "action_mask"
+        )
+
+        agents_view = specs.BoundedArray(
+            (self.num_agents, (self.num_customers + 1) * 7 + 4),
+            jnp.float32,
+            -jnp.inf,
+            jnp.inf,
+            "agents_view",
+        )
+        obs_data = {
+            "agents_view": agents_view,
+            "action_mask": action_mask,
+            "step_count": step_count,
+        }
+
+        if self.has_global_state:
+            global_state = specs.Array(
+                (self.num_agents, (self.num_customers + 1) * 7 + 4 * self.num_agents),
+                jnp.float32,
+                "global_state",
+            )
+            obs_data["global_state"] = global_state
+            return specs.Spec(ObservationGlobalState, "ObservationSpec", **obs_data)
+
+        return specs.Spec(Observation, "ObservationSpec", **obs_data)
+
+    def action_spec(self) -> specs.Spec:
+        return specs.MultiDiscreteArray(
+            num_values=jnp.full(self.num_agents, self.num_customers + 1)
+        )
```

Comment:
I didn't pay attention to this last time but can you please add a comment here to clarify why you added this `action_spec `and also I would suggest overriding the `action_dim` method instead of setting it manually and this helps to link things better and use this `action_spec` instead of the original 🙏

Reply:
Turns out the base `action_spec` is already sufficient!